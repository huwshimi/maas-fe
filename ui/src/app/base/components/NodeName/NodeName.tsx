import { useEffect, useState } from "react";

import { Button, Spinner } from "@canonical/react-components";
import { usePrevious } from "@canonical/react-components/dist/hooks";
import type { FormikErrors } from "formik";
import { Formik } from "formik";
import * as Yup from "yup";

import NodeNameFields from "./NodeNameFields";

import FormikFormContent from "app/base/components/FormikFormContent";
import { useCanEdit } from "app/base/hooks";
import type { Device } from "app/store/device/types";
import type { Domain, DomainMeta } from "app/store/domain/types";
import type { Machine } from "app/store/machine/types";
import type { SimpleNode } from "app/store/types/node";

export type Props = {
  editingName: boolean;
  // Machines and devices can edit their name, but no controllers.
  node: Machine | Device | null;
  onSubmit: (
    hostname: SimpleNode["hostname"],
    domain: Domain[DomainMeta.PK]
  ) => void;
  setEditingName: (editingName: boolean) => void;
  saved?: boolean;
  saving?: boolean;
};

export type FormValues = {
  hostname: string;
  domain: string;
};

const hostnamePattern = /^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])*$/;

export enum Label {
  HostnamePatternError = "Hostname must only contain letters, numbers and hyphens.",
}

const Schema = Yup.object().shape({
  hostname: Yup.string()
    .max(63, "Hostname must be 63 characters or less.")
    .matches(hostnamePattern, Label.HostnamePatternError)
    .required(),
  domain: Yup.string(),
});

const NodeName = ({
  editingName,
  node,
  onSubmit,
  setEditingName,
  saved,
  saving,
}: Props): JSX.Element => {
  const [hostnameError, setHostnameError] = useState<
    FormikErrors<FormValues>["hostname"] | null
  >(null);
  const canEdit = useCanEdit(node);
  const previousSaving = usePrevious(saving);

  useEffect(() => {
    // The node has transitioned from saving to saved so close the form.
    if (saved && !saving && previousSaving) {
      setEditingName(false);
    }
  }, [previousSaving, saved, saving, setEditingName]);

  if (!node) {
    return <Spinner />;
  }
  if (!canEdit) {
    return <span className="node-name">{node.fqdn}</span>;
  }
  if (!editingName) {
    return (
      <Button
        appearance="base"
        className="node-name--editable"
        onClick={() => setEditingName(true)}
      >
        {node.fqdn}
      </Button>
    );
  }
  return (
    <Formik
      initialValues={{
        domain: String(node.domain.id),
        hostname: node.hostname,
      }}
      onSubmit={({ hostname, domain }) => {
        onSubmit(hostname, Number(domain));
      }}
      validationSchema={Schema}
    >
      <FormikFormContent<FormValues>
        buttonsAlign="right"
        buttonsBordered={false}
        className="node-name"
        footer={
          hostnameError ? (
            <div className="node-name__error is-error">
              <p className="p-form-validation__message u-no-margin--bottom">
                <strong>Error:</strong> {hostnameError}
              </p>
            </div>
          ) : null
        }
        inline
        onCancel={() => setEditingName(false)}
        onSaveAnalytics={{
          action: "Saved",
          category: "Node details header",
          label: "name",
        }}
        saving={saving}
        saved={saved}
      >
        <NodeNameFields saving={saving} setHostnameError={setHostnameError} />
      </FormikFormContent>
    </Formik>
  );
};

export default NodeName;
