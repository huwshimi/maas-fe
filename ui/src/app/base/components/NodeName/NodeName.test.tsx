import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";

import NodeName from "./NodeName";
import type { Props as NodeNameProps } from "./NodeName";
import NodeNameFields from "./NodeNameFields";

import type { Machine } from "app/store/machine/types";
import type { RootState } from "app/store/root/types";
import {
  domain as domainFactory,
  domainState as domainStateFactory,
  generalState as generalStateFactory,
  machineDetails as machineDetailsFactory,
  machineState as machineStateFactory,
  powerType as powerTypeFactory,
  powerTypesState as powerTypesStateFactory,
  rootState as rootStateFactory,
} from "testing/factories";

const mockStore = configureStore();

describe("NodeName", () => {
  let state: RootState;
  let machine: Machine;

  beforeEach(() => {
    const domain = domainFactory({ id: 99 });
    machine = machineDetailsFactory({
      domain,
      locked: false,
      permissions: ["edit"],
      system_id: "abc123",
    });
    state = rootStateFactory({
      domain: domainStateFactory({
        items: [domain],
      }),
      general: generalStateFactory({
        powerTypes: powerTypesStateFactory({
          data: [powerTypeFactory()],
        }),
      }),
      machine: machineStateFactory({
        loaded: true,
        items: [machine],
      }),
    });
  });

  it("displays a spinner when loading", () => {
    state.machine.items = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName
            editingName={false}
            node={null}
            onSubmit={jest.fn()}
            setEditingName={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Spinner").exists()).toBe(true);
  });

  it("displays just the name when not editable", () => {
    state.machine.items[0].locked = true;
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName
            editingName={false}
            node={machine}
            onSubmit={jest.fn()}
            setEditingName={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".node-name").exists()).toBe(true);
  });

  it("displays name in a button", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName
            editingName={false}
            node={machine}
            onSubmit={jest.fn()}
            setEditingName={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Button.node-name--editable").exists()).toBe(true);
  });

  it("changes the form state when clicking the name", () => {
    const setEditingName = jest.fn();
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName
            editingName={false}
            node={machine}
            onSubmit={jest.fn()}
            setEditingName={setEditingName}
          />
        </MemoryRouter>
      </Provider>
    );
    wrapper.find("Button.node-name--editable").simulate("click");
    expect(setEditingName).toHaveBeenCalled();
  });

  it("can display the form", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName
            editingName={true}
            node={machine}
            onSubmit={jest.fn()}
            setEditingName={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("FormikFormContent")).toMatchSnapshot();
  });

  it("closes the form when it saves", () => {
    state.machine.saving = true;
    const setEditingName = jest.fn();
    const store = mockStore(state);
    const ProxyNodeName = (props: NodeNameProps) => (
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName {...props} />
        </MemoryRouter>
      </Provider>
    );
    const wrapper = mount(
      <ProxyNodeName
        editingName={true}
        node={machine}
        onSubmit={jest.fn()}
        setEditingName={setEditingName}
        saved={false}
        saving={true}
      />
    );
    wrapper.setProps({
      saved: true,
      saving: false,
    });
    expect(setEditingName).toHaveBeenCalledWith(false);
  });

  it("can display a hostname error", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <NodeName
            editingName={true}
            node={machine}
            onSubmit={jest.fn()}
            setEditingName={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    act(() => {
      wrapper.find(NodeNameFields).props().setHostnameError("Uh oh!");
    });
    wrapper.update();
    const error = wrapper.find(".node-name__error");
    expect(error.exists()).toBe(true);
    expect(error.text()).toBe("Error: Uh oh!");
  });
});
