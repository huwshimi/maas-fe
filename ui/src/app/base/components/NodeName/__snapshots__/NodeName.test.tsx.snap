// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeName can display the form 1`] = `
<FormikFormContent
  buttonsAlign="right"
  buttonsBordered={false}
  className="node-name"
  footer={null}
  inline={true}
  onCancel={[Function]}
  onSaveAnalytics={
    Object {
      "action": "Saved",
      "category": "Node details header",
      "label": "name",
    }
  }
>
  <Form
    className="node-name"
    inline={true}
    onSubmit={[Function]}
  >
    <form
      className="node-name p-form p-form--inline"
      onSubmit={[Function]}
    >
      <NodeNameFields
        setHostnameError={[Function]}
      >
        <div
          className="node-name__hostname-wrapper u-no-margin--right"
        >
          <div
            aria-hidden="true"
            className="node-name__hostname-spacer"
          >
            test-machine-5
          </div>
          <FormikField
            className="node-name__hostname"
            displayError={false}
            label="Hostname"
            name="hostname"
            takeFocus={true}
            type="text"
            wrapperClassName="u-no-margin--right"
          >
            <Input
              aria-label="Hostname"
              className="node-name__hostname"
              error={null}
              id="mock-redux-js-nanoid-1"
              label="Hostname"
              name="hostname"
              onBlur={[Function]}
              onChange={[Function]}
              takeFocus={true}
              type="text"
              value="test-machine-5"
              wrapperClassName="u-no-margin--right"
            >
              <Field
                className="u-no-margin--right"
                error={null}
                forId="mock-redux-js-nanoid-1"
                helpId="mock-nanoid-2"
                isTickElement={false}
                label="Hostname"
                validationId="mock-nanoid-1"
              >
                <div
                  className="p-form__group p-form-validation u-no-margin--right"
                >
                  <Label
                    forId="mock-redux-js-nanoid-1"
                  >
                    <label
                      className="p-form__label"
                      htmlFor="mock-redux-js-nanoid-1"
                    >
                      Hostname
                    </label>
                  </Label>
                  <div
                    className="p-form__control u-clearfix"
                  >
                    <input
                      aria-describedby={null}
                      aria-errormessage={null}
                      aria-invalid={false}
                      aria-label="Hostname"
                      className="p-form-validation__input node-name__hostname"
                      id="mock-redux-js-nanoid-1"
                      label="Hostname"
                      name="hostname"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="text"
                      value="test-machine-5"
                    />
                  </div>
                </div>
              </Field>
            </Input>
          </FormikField>
        </div>
        <span
          className="node-name__dot u-nudge-right--small u-nudge-left--small u-no-margin--right"
        >
          .
        </span>
        <Spinner
          className="u-width--auto"
        >
          <span
            aria-live="polite"
            className="u-width--auto p-text--default"
            role="alert"
          >
            <i
              className="p-icon--spinner u-animation--spin"
            >
              Loading
            </i>
          </span>
        </Spinner>
      </NodeNameFields>
      <FormikFormButtons
        buttonsAlign="right"
        buttonsBordered={false}
        inline={true}
        onCancel={[Function]}
        saved={false}
        submitDisabled={true}
      >
        <div
          className="formik-form-buttons is-inline"
          data-testid="buttons-wrapper"
        >
          <div
            className="formik-form-buttons__container u-align--right"
          >
            <Button
              appearance="base"
              className="formik-form-buttons__button"
              data-testid="cancel-action"
              onClick={[Function]}
              type="button"
            >
              <button
                className="p-button--base formik-form-buttons__button"
                data-testid="cancel-action"
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
            </Button>
            <ActionButton
              appearance="positive"
              className="formik-form-buttons__button"
              disabled={true}
              success={false}
              type="submit"
            >
              <button
                className="formik-form-buttons__button p-action-button p-button--positive is-disabled"
                disabled={true}
                type="submit"
              >
                Save
              </button>
            </ActionButton>
          </div>
        </div>
      </FormikFormButtons>
    </form>
  </Form>
</FormikFormContent>
`;
